### PR TITLE
Add logs to VMware instance configuration process

### PR DESCRIPTION
**Pull request by winterhazel**
### Description

The virtual machine configuration process is used in a lot of VMware workflows, such as instance creation, start and scale. However, there are no logs in this process, which difficults the troubleshooting of errors, specially of those that are returned by VMware.

This PR adds logs to this process in order to make the troubleshooting of errors easier.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [X] Minor

### How Has This Been Tested?

In a local lab:

1. I created a VM using only IDE controllers;
2. I stopped the VM;
3. I changed the controllers of the VM from IDE to SCSI;
4. I started the VM, and verified that the logs presented the reconfigurations of the VM in order to remove the disks and NICs, to add the SCSI controllers, and the general VM configuration.

```bash
ubuntu@lab-vmware-mgmt-01:~$ sudo grep 'Reconfiguring virtual machine' $mgmt_log
2024-05-14T11:17:15,122 DEBUG [c.c.h.v.m.VirtualMachineMO] (DirectAgent-147:[ctx-9fd2c09e, 172.16.75.21, job-783/job-784, cmd: StartCommand]) (logid:64cbc567) Reconfiguring virtual machine [{"internalCSName":"i-2-217-VM"}] using spec [{"deviceChange":[{"operation":"REMOVE","device":{"capacityInKB":51200,"capacityInBytes":52428800,"shares":{"shares":1000,"level":"NORMAL"},"storageIOAllocation":{"limit":-1,"shares":{"shares":1000,"level":"NORMAL"},"reservation":0},"diskObjectId":"518-3001","key":3001,"deviceInfo":{"label":"Hard disk 1","summary":"51,200 KB"},"backing":{"diskMode":"persistent","split":false,"writeThrough":false,"thinProvisioned":true,"uuid":"6000C29b-5522-5947-c391-e29f44c11a73","contentId":"fb640b892cf0174da714e57339203aa5","digestEnabled":false,"sharing":"sharingNone","fileName":"[c2b40d03a7b132fc928f7f9f39cf4297] i-2-217-VM/ROOT-217.vmdk","datastore":{"value":"datastore-22","type":"Datastore"},"backingObjectId":""},"controllerKey":200,"unitNumber":1}},{"operation":"REMOVE","device":{"addressType":"manual","macAddress":"02:01:00:d0:00:04","wakeOnLanEnabled":true,"resourceAllocation":{"reservation":0,"share":{"shares":50,"level":"NORMAL"},"limit":-1},"uptCompatibilityEnabled":false,"key":4000,"deviceInfo":{"label":"Network adapter 1","summary":"cloud.guest.615.200.1-vSwitch0"},"backing":{"network":{"value":"network-380","type":"Network"},"deviceName":"cloud.guest.615.200.1-vSwitch0","useAutoDetect":false},"connectable":{"startConnected":true,"allowGuestControl":true,"connected":false,"status":"untried"},"slotInfo":{"pciSlotNumber":32},"controllerKey":100,"unitNumber":7}}]}].
2024-05-14T11:17:16,964 DEBUG [c.c.h.v.m.VirtualMachineMO] (DirectAgent-147:[ctx-9fd2c09e, 172.16.75.21, job-783/job-784, cmd: StartCommand]) (logid:64cbc567) Reconfiguring virtual machine [{"internalCSName":"i-2-217-VM"}] using spec [{"deviceChange":[{"operation":"ADD","device":{"sharedBus":"NO_SHARING","busNumber":0,"key":-4}},{"operation":"ADD","device":{"sharedBus":"NO_SHARING","busNumber":1,"key":-3}},{"operation":"ADD","device":{"sharedBus":"NO_SHARING","busNumber":2,"key":-2}},{"operation":"ADD","device":{"sharedBus":"NO_SHARING","busNumber":3,"key":-1}}]}].
2024-05-14T11:17:29,603 DEBUG [c.c.h.v.m.VirtualMachineMO] (DirectAgent-149:[ctx-a1985c08, 172.16.75.21, job-783/job-785, cmd: StartCommand]) (logid:64cbc567) Reconfiguring virtual machine [{"internalCSName":"i-2-217-VM"}] using spec [{"guestId":"otherLinux64Guest","numCPUs":1,"memoryMB":512,"memoryHotAddEnabled":false,"cpuHotAddEnabled":false,"deviceChange":[{"operation":"EDIT","device":{"key":3000,"deviceInfo":{"label":"CD/DVD drive 1","summary":"Remote device"},"backing":{"exclusive":false,"deviceName":""},"connectable":{"startConnected":true,"allowGuestControl":false,"connected":true},"controllerKey":200,"unitNumber":0}},{"operation":"ADD","device":{"capacityInKB":0,"key":-2,"backing":{"diskMode":"persistent","fileName":"[c2b40d03a7b132fc928f7f9f39cf4297] i-2-217-VM/ROOT-217.vmdk","datastore":{"value":"datastore-22","type":"Datastore"}},"connectable":{"startConnected":true,"allowGuestControl":false,"connected":true},"controllerKey":1000,"unitNumber":0}},{"operation":"ADD","device":{"addressType":"Manual","macAddress":"02:01:00:d0:00:04","key":-3,"backing":{"network":{"value":"network-380","type":"Network"},"deviceName":"cloud.guest.615.200.1-vSwitch0"},"connectable":{"startConnected":true,"allowGuestControl":true,"connected":true}}}],"cpuAllocation":{"reservation":0,"limit":-1},"memoryAllocation":{"reservation":0,"limit":512},"extraConfig":[{"key":"machine.id","value":""},{"key":"devices.hotplug","value":"true"},{"key":"nvp.vm-uuid","value":"ee781edb-bb7b-4d15-8dfe-88b582bc1419"},{"key":"nvp.iface-id.0","value":"47973c05-5956-4c89-a0d8-b7ef218cb7de"},{"key":"dataDiskController","value":"scsi"},{"key":"Message.ReservedCapacityFreed.Flag","value":"false"},{"key":"nestedVirtualizationFlag","value":"false"},{"key":"vmware.reserve.mem","value":"false"},{"key":"vmware.reserve.cpu","value":"false"},{"key":"nicAdapter","value":"E1000"},{"key":"rootDiskController","value":"scsi"},{"key":"RemoteDisplay.vnc.enabled","value":"true"},{"key":"RemoteDisplay.vnc.password","value":"xDys93MKuuAJTnvoPya1tw"},{"key":"RemoteDisplay.vnc.port","value":"5918"}]}].
```

### CODE REVIEW

File: vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/VirtualMachineMO.java [hunk 1]
```diff
@@ -1182,6 +1182,7 @@ public boolean setVncConfigInfo(boolean enableVnc, String vncPassword, int vncPo
     }
 
     public boolean configureVm(VirtualMachineConfigSpec vmConfigSpec) throws Exception {
+        logger.debug(String.format("Reconfiguring virtual machine [%s] using spec [%s].", this, GSON.toJson(vmConfigSpec)));
```
Comment (hsato03) [inline vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/VirtualMachineMO.java:1185]:
```suggestion
        logger.debug("Reconfiguring virtual machine {} using spec {}.", this, GSON.toJson(vmConfigSpec));
```

**Review by hsato03 — APPROVED** [review]
LGTM

**Comment by winterhazel** [comment]
@blueorangutan package

**Review by DaanHoogland — APPROVED** [review]
clgtm in principal. Did you check if this object is not `toString()`d anywhere else, @winterhazel ?

**Comment by DaanHoogland** [comment]
Did you test @hsato03 ?

**Comment by DaanHoogland** [comment]
> > clgtm in principal. Did you check if this object is not `toString()`d anywhere else, @winterhazel ?
> 
> @DaanHoogland yes, its `toString()` is not used anywhere, as its not even implemented. Whenever we need to log this object, we're converting it to JSON.

ok, so no default or inherrited toString used anywhere? great.

**Comment by JoaoJandre** [comment]
@blueorangutan package

**Comment by winterhazel** [comment]
@DaanHoogland @rohityadavcloud can we run the CI?

**Review by kiranchavala — APPROVED** [review]
LGTM

Followed the steps 


    I created a VM using only IDE controllers;
    I stopped the VM;
    I changed the controllers of the VM from IDE to SCSI;
    I started the VM, and verified that the logs presented the reconfigurations of the VM in order to remove the disks and NICs, to add the SCSI controllers, and the general VM configuration.


The log message contains the reconfiguration message 

```
2024-09-24T04:53:44,534 DEBUG [c.c.h.v.m.VirtualMachineMO] (DirectAgent-155:[ctx-398bbb74, 10.0.34.240, job-38/job-39, cmd: StartCommand]) (logid:409ef4d5) Reconfiguring virtual machine {"internalCSName":"i-2-3-VM"} using spec {"guestId":"centos64Guest","numCPUs":1,"memoryMB":512,"memoryHotAddEnabled":false,"cpuHotAddEnabled":false,"deviceChange":[{"operation":"ADD","device":{"key":-1,"backing":{"exclusive":false,"deviceName":""},"connectable":{"startConnected":true,"allowGuestControl":false,"connected":true},"controllerKey":200,"unitNumber":0}},{"operation":"ADD","device":{"capacityInKB":0,"key":-2,"backing":{"diskMode":"persistent","fileName":"[afb511315e833dfe933fff7fb5d4f735] i-2-3-VM/ROOT-3.vmdk","datastore":{"value":"datastore-1013","type":"Datastore"}},"connectable":{"startConnected":true,"allowGuestControl":false,"connected":true},"controllerKey":200,"unitNumber":1}},{"operation":"ADD","device":{"addressType":"Manual","macAddress":"02:01:00:cc:00:01","key":-3,"backing":{"network":{"value":"network-1025","type":"Network"},"deviceName":"cloud.guest.1315.200.1-vSwitch1"},"connectable":{"startConnected":true,"allowGuestControl":true,"connected":true}}}],"cpuAllocation":{"reservation":0,"limit":-1},"memoryAllocation":{"reservation":0,"limit":512},"extraConfig":[{"key":"machine.id","value":""},{"key":"devices.hotplug","value":"true"},{"key":"nvp.vm-uuid","value":"4afa4e00-e824-43ea-9372-2dcd166aac5f"},{"key":"nvp.iface-id.0","value":"4bd87b82-3b90-424e-a893-86fc9a47f78b"},{"key":"deployvm","value":"true"},{"key":"dataDiskController","value":"osdefault"},{"key":"nestedVirtualizationFlag","value":"false"},{"key":"cpuOvercommitRatio","value":"2.0"},{"key":"vmware.reserve.mem","value":"false"},{"key":"vmware.reserve.cpu","value":"false"},{"key":"nicAdapter","value":"E1000"},{"key":"rootDiskController","value":"ide"},{"key":"RemoteDisplay.vnc.enabled","value":"true"},{"key":"RemoteDisplay.vnc.password","value":"UYstXkIEeJstmmgG4q7bKQ"},{"key":"RemoteDisplay.vnc.port","value":"5942"}]}.



2024-09-24T04:56:43,698 DEBUG [c.c.h.v.m.VirtualMachineMO] (DirectAgent-318:[ctx-d97a7f50, 10.0.34.240, job-43/job-44, cmd: StartCommand]) (logid:827dbdf7) Reconfiguring virtual machine {"internalCSName":"i-2-3-VM"} using spec {"deviceChange":[{"operation":"REMOVE","device":{"capacityInKB":2097152,"capacityInBytes":2147483648,"shares":{"shares":1000,"level":"NORMAL"},"storageIOAllocation":{"limit":-1,"shares":{"shares":1000,"level":"NORMAL"},"reservation":0},"diskObjectId":"11-3001","nativeUnmanagedLinkedClone":false,"key":3001,"deviceInfo":{"label":"Hard disk 1","summary":"2,097,152 KB"},"backing":{"diskMode":"persistent","split":false,"writeThrough":false,"thinProvisioned":true,"eagerlyScrub":false,"uuid":"6000C29d-2763-0bb8-e26b-39fbd7075417","contentId":"25f1592693ef16a6555911cdcd4465a3","parent":{"diskMode":"persistent","thinProvisioned":true,"eagerlyScrub":false,"uuid":"6000C29d-2763-0bb8-e26b-39fbd7075417","contentId":"25f1592693ef16a6555911cdcd4465a3","digestEnabled":false,"fileName":"[afb511315e833dfe933fff7fb5d4f735] 1dbafacae14231779a8b33a4d0dd0d99/1dbafacae14231779a8b33a4d0dd0d99.vmdk","datastore":{"value":"datastore-1013","type":"Datastore"},"backingObjectId":""},"deltaDiskFormat":"redoLogFormat","digestEnabled":false,"deltaDiskFormatVariant":"vmfsSparseVariant","sharing":"sharingNone","fileName":"[afb511315e833dfe933fff7fb5d4f735] i-2-3-VM/ROOT-3.vmdk","datastore":{"value":"datastore-1013","type":"Datastore"},"backingObjectId":""},"controllerKey":200,"unitNumber":1}},{"operation":"REMOVE","device":{"addressType":"manual","macAddress":"02:01:00:cc:00:01","wakeOnLanEnabled":true,"resourceAllocation":{"reservation":0,"share":{"shares":50,"level":"NORMAL"},"limit":-1},"uptCompatibilityEnabled":false,"key":4000,"deviceInfo":{"label":"Network adapter 1","summary":"cloud.guest.1315.200.1-vSwitch1"},"backing":{"network":{"value":"network-1025","type":"Network"},"deviceName":"cloud.guest.1315.200.1-vSwitch1","useAutoDetect":false},"connectable":{"migrateConnect":"unset","startConnected":true,"allowGuestControl":true,"connected":false,"status":"untried"},"slotInfo":{"pciSlotNumber":35},"controllerKey":100,"unitNumber":7}}]}.



2024-09-24T04:56:44,589 DEBUG [c.c.h.v.m.VirtualMachineMO] (DirectAgent-318:[ctx-d97a7f50, 10.0.34.240, job-43/job-44, cmd: StartCommand]) (logid:827dbdf7) Reconfiguring virtual machine {"internalCSName":"i-2-3-VM"} using spec {"guestId":"centos64Guest","numCPUs":1,"memoryMB":512,"memoryHotAddEnabled":false,"cpuHotAddEnabled":false,"deviceChange":[{"operation":"EDIT","device":{"key":3000,"deviceInfo":{"label":"CD/DVD drive 1","summary":"Remote device"},"backing":{"exclusive":false,"deviceName":""},"connectable":{"startConnected":true,"allowGuestControl":false,"connected":true},"controllerKey":200,"unitNumber":0}},{"operation":"ADD","device":{"capacityInKB":0,"key":-2,"backing":{"diskMode":"persistent","parent":{"diskMode":"persistent","fileName":"[afb511315e833dfe933fff7fb5d4f735] 1dbafacae14231779a8b33a4d0dd0d99/1dbafacae14231779a8b33a4d0dd0d99.vmdk","datastore":{"value":"datastore-1013","type":"Datastore"}},"fileName":"[afb511315e833dfe933fff7fb5d4f735] i-2-3-VM/ROOT-3.vmdk","datastore":{"value":"datastore-1013","type":"Datastore"}},"connectable":{"startConnected":true,"allowGuestControl":false,"connected":true},"controllerKey":1000,"unitNumber":0}},{"operation":"ADD","device":{"addressType":"Manual","macAddress":"02:01:00:cc:00:01","key":-3,"backing":{"network":{"value":"network-1025","type":"Network"},"deviceName":"cloud.guest.1315.200.1-vSwitch1"},"connectable":{"startConnected":true,"allowGuestControl":true,"connected":true}}}],"cpuAllocation":{"reservation":0,"limit":-1},"memoryAllocation":{"reservation":0,"limit":512},"extraConfig":[{"key":"machine.id","value":""},{"key":"devices.hotplug","value":"true"},{"key":"nvp.vm-uuid","value":"4afa4e00-e824-43ea-9372-2dcd166aac5f"},{"key":"nvp.iface-id.0","value":"4bd87b82-3b90-424e-a893-86fc9a47f78b"},{"key":"dataDiskController","value":"osdefault"},{"key":"Message.ReservedCapacityFreed.Flag","value":"false"},{"key":"nestedVirtualizationFlag","value":"false"},{"key":"cpuOvercommitRatio","value":"2.0"},{"key":"vmware.reserve.mem","value":"false"},{"key":"vmware.reserve.cpu","value":"false"},{"key":"nicAdapter","value":"E1000"},{"key":"rootDiskController","value":"scsi"},{"key":"RemoteDisplay.vnc.enabled","value":"true"},{"key":"RemoteDisplay.vnc.password","value":"UYstXkIEeJstmmgG4q7bKQ"},{"key":"RemoteDisplay.vnc.port","value":"5928"}]}.
```

**Pull request merged**